File: app/stardag-ui/src/utils/deployments.ts
import type { Deployment, DeploymentKind } from "../types/task";

/** One app's deployments: every generation, newest first. */
export interface AppDeployments {
  kind: DeploymentKind;
  appName: string;
  current: Deployment | null;
  generations: Deployment[];
}

/**
 * Group deployments by `(kind, app_name)` — generations are monotonic per
 * that pair, not per app name alone, so a local and a Modal app of one
 * name are two groups. Groups are sorted by app name, then kind.
 */
export function groupByApp(deployments: Deployment[]): AppDeployments[] {
  const groups = new Map<string, AppDeployments>();
  for (const deployment of deployments) {
    const key = `${deployment.kind}\u0000${deployment.app_name}`;
    let group = groups.get(key);
    if (!group) {
      group = {
        kind: deployment.kind,
        appName: deployment.app_name,
        current: null,
        generations: [],
      };
      groups.set(key, group);
    }
    group.generations.push(deployment);
    if (deployment.is_current) group.current = deployment;
  }
  const result = [...groups.values()];
  for (const group of result) {
    group.generations.sort((a, b) => b.generation - a.generation);
  }
  return result.sort(
    (a, b) => a.appName.localeCompare(b.appName) || a.kind.localeCompare(b.kind),
  );
}

/** "app gen 3", the short form a plan names its deployment by. */
export function deploymentLabel(deployment: Deployment): string {
  return `${deployment.app_name} gen ${deployment.generation}`;
}

File: lib/stardag/src/stardag/_cli/deployments.py
"""``stardag deployments``: the deployments recorded in the registry (v2).

One row per ``stardag modal deploy`` (``kind=modal``), plus one per local
code id a local build planned under (``kind=local``). An app's **current**
deployment is its activated Modal row with the highest generation; local
rows are never current (design.md, "The deterministic scope").
"""

from typing import Optional, cast

import typer
from rich.table import Table

from stardag._cli._output import JSON_OPTION, emit_json, short, stamp
from stardag._cli._registry_ctx import (
    _ENV_OPTION,
    _PROFILE_OPTION,
    _fail,
    _resolve_registry,
    console,
    error_console,
)
from stardag.exceptions import StardagError
from stardag.registry import DeploymentInfo
from stardag.registry._models import DeploymentKind

app = typer.Typer(help="Inspect recorded deployments.", no_args_is_help=True)


def check_kind(kind: Optional[str]) -> None:
    if kind not in (None, "modal", "local"):
        error_console.print(
            f"[bold red]Error:[/bold red] --kind is 'modal' or 'local', got {kind!r}."
        )
        raise typer.Exit(1)


def render_deployments(rows: list[DeploymentInfo], *, json_output: bool) -> None:
    """The listing, shared with ``stardag modal deployments``."""
    if json_output:
        emit_json({"deployments": [d.model_dump(mode="json") for d in rows]})
        return
    if not rows:
        console.print("No deployments recorded.")
        return
    table = Table(title="Deployments (newest first)")
    for col in (
        "Kind",
        "App",
        "Deployment",
        "Gen",
        "Code id",
        "Deployed",
        "Activated",
        "",
    ):
        table.add_column(col)
    for d in rows:
        if d.is_current:
            mark = "[green]current[/green]"
        elif d.kind == "modal" and d.activated_at is None:
            mark = "[yellow]not activated[/yellow]"
        else:
            mark = ""
        table.add_row(
            d.kind,
            d.app_name,
            str(d.id),
            str(d.generation),
            short(d.code_id),
            stamp(d.deployed_at),
            stamp(d.activated_at),
            mark,
        )
    console.print(table)


@app.command("list")
def deployments_list(
    app_name: Optional[str] = typer.Option(None, "--app", help="Only this app."),
    kind: Optional[str] = typer.Option(
        None, "--kind", help="Only 'modal' or only 'local' deployments."
    ),
    current: bool = typer.Option(
        False, "--current", help="Only each app's current deployment."
    ),
    limit: int = typer.Option(100, "--limit", "-n", min=1, max=500),
    stardag_profile: Optional[str] = _PROFILE_OPTION,
    stardag_env: Optional[str] = _ENV_OPTION,
    json_output: bool = JSON_OPTION,
) -> None:
    """List deployments, newest first, marking each app's current one.

    Reads ``GET /deployments``. Writes nothing. A Modal row that is not
    activated is a deploy whose record was created but whose activation
    never landed: until it does, the new code's ticks exit ``superseded``
    (re-run ``stardag modal deploy``).
    """
    check_kind(kind)
    registry = _resolve_registry(stardag_profile, stardag_env)
    try:
        rows = registry.deployment_list(
            kind=cast(Optional[DeploymentKind], kind),
            app_name=app_name,
            current=current,
            limit=limit,
        )
    except StardagError as e:
        _fail(e)
    finally:
        registry.close()
    render_deployments(rows, json_output=json_output)
